federation/pkg/federation-controller/replicaset/planner/planer.go

```diff
@@ -0,0 +1,144 @@
+
+func (a byWeight) Len() int      { return len(a) }
+func (a byWeight) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+// Preferences are sorted according by decreasing weight and increasing clusterName.
+func (a byWeight) Less(i, j int) bool {
+	return (a[i].Weight > a[j].Weight) || (a[i].Weight == a[j].Weight && a[i].clusterName < a[j].clusterName)
+}
+
+func NewPlaner(preferences *fed_api.FederatedReplicaSetPreferences) *Planer {
+	return &Planer{
+		preferences: preferences,
+	}
+}
+
+// Distribute the desired number of replicas among the given cluster according to the planner preferences.
+func (p *Planer) Plan(replicasToDistribute int64, availableClusters []string) map[string]int64 {
+	preferences := make([]*namedLocalReplicaSetPreferences, 0, len(availableClusters))
+
+	named := func(name string, pref fed_api.LocalReplicaSetPreferences) *namedLocalReplicaSetPreferences {
+		return &namedLocalReplicaSetPreferences{
+			clusterName:                name,
+			LocalReplicaSetPreferences: pref,
+		}
+	}
+
+	for _, cluster := range availableClusters {
+		if localRSP, found := p.preferences.Clusters[cluster]; found {
+			preferences = append(preferences, named(cluster, localRSP))
+		} else {
+			if localRSP, found := p.preferences.Clusters["*"]; found {
+				preferences = append(preferences, named(cluster, localRSP))
+			} else {
+				preferences = append(preferences, named(cluster, fed_api.LocalReplicaSetPreferences{Weight: 0}))
+			}
+		}
+	}
+	sort.Sort(byWeight(preferences))
+
+	remainingReplicas := replicasToDistribute
+	plan := make(map[string]int64, len(preferences))
+
+	// Assign each cluster the minimum number replica it requested.
+	for _, preference := range preferences {
+		min := minInt64(preference.MinReplicas, remainingReplicas)
+		remainingReplicas -= min
+		plan[preference.clusterName] = min
+	}
+
+	modified := true
+
+	// It is possible single pass of the loop is not enough to distribue all replicas among clusters due
+	// to weight, max and rounding corner cases. In such case we iterate until either
+	// there is no replicas or no cluster gets any more replicas or the number
+	// of attempts is less than available couster count. Every loop either distributes all remainingReplicas
+	// or maxes out at least one cluster.
+	// TODO: This algorithm is O(clusterCount^2). Find something faster.
+	for trial := 0; trial < len(availableClusters) && modified && remainingReplicas > 0; trial++ {
+		modified = false
+		weightSum := int64(0)
+		for _, preference := range preferences {
+			weightSum += preference.Weight
+		}
+		newPreferences := make([]*namedLocalReplicaSetPreferences, 0, len(preferences))
+
+		distributeInThisLoop := remainingReplicas
+		for _, preference := range preferences {
+			if weightSum > 0 {
+				start := plan[preference.clusterName]
+				// Distribute the remaining replicas, rounding fractions up from .5 and down from .5
+				extra := (distributeInThisLoop*preference.Weight + weightSum/int64(2)) / weightSum
```

Comment:
Yes I want to (calculate distributeInThisLoop * preference.Weight/weightSum) but without floats. There was an issue with the formula but a different one with the idea of rounding. It should actually be ceil. Rounding is wrong. Imagine distributing 2 replicas among 5 clusters. Each of them gets 0.4 of replica which is rounded to 0. Its better to round up. Then two first will get 1 replica and other, who would also like a replica wound't get any because there will be no replicas let.